### ISSUE TITLE
TEST-07: Optimize PostgreSQL test configuration

### ISSUE THREAD

**Issue by DegrassiAaron** [DegrassiAaron/meepleai-monorepo#615]
## Objective
Configure PostgreSQL for test environments with relaxed durability settings to achieve 30-50% faster database operations.

## Background
Production PostgreSQL prioritizes data durability with fsync, synchronous commits, and full page writes. Test environments can safely disable these for significant performance gains.

**Research Reference**: `claudedocs/research_testcontainers_optimization_2025-11-01.md` Section 4.2, 4.3

⚠️ **WARNING**: These settings are UNSAFE for production. Use ONLY in test environments.

### 1. Add PostgreSQL Test Configuration (30min)
Update `IntegrationTestFixture`:
```csharp
private async Task StartPostgresAsync()
{
    PostgresContainer = new PostgreSqlBuilder()
        .WithImage("postgres:15-alpine")
        .WithDatabase("meepleai_test")
        .WithUsername("postgres")
        .WithPassword("testpassword")
        .WithCommand(
            "postgres",
            // Disable fsync - Don't wait for disk writes
            "-c", "fsync=off",
            // Disable synchronous commits - Don't wait for WAL writes
            "-c", "synchronous_commit=off",
            // Disable full page writes - Reduce WAL size
            "-c", "full_page_writes=off",
            // Increase connection pool for parallel tests
            "-c", "max_connections=500",
            // Increase shared buffers for better caching
            "-c", "shared_buffers=256MB",
            // Disable background writer for tests
            "-c", "bgwriter_lru_maxpages=0"
        )
        .Build();

    await PostgresContainer.StartAsync();
}
```

### 2. Add tmpfs Volume for Database Storage (1h)
Use RAM disk instead of disk I/O for maximum speed:
```csharp
.WithResourceMapping(new MountConfiguration
{
    Type = MountType.Tmpfs,
    Target = "/var/lib/postgresql/data",
    TmpfsOptions = new TmpfsOptions 
    { 
        SizeBytes = 1073741824,  // 1GB RAM disk
        Mode = "1777"
    }
})
```

### 3. Document Safety Guardrails (30min)
Add comments and runtime checks:
```csharp
// ⚠️ WARNING: These settings DISABLE PostgreSQL durability guarantees
// NEVER use these settings in production environments
// Safe for tests because:
// 1. Test data is ephemeral (containers destroyed after tests)
// 2. Test failures are acceptable (tests can be re-run)
// 3. Data loss on crash is acceptable (no real data)

#if !DEBUG
#error "PostgreSQL test configuration must not be used in production builds"
#endif
```

### 4. Benchmark Performance Improvement (1h)
Create benchmark test:
```csharp
[Fact]
public async Task Benchmark_InsertPerformance()
{
    var stopwatch = Stopwatch.StartNew();
    
    for (int i = 0; i < 1000; i++)
    {
        DbContext.Users.Add(new User { Email = $"user{i}@test.com" });
    }
    await DbContext.SaveChangesAsync();
    
    stopwatch.Stop();
    _output.WriteLine($"1000 inserts: {stopwatch.ElapsedMilliseconds}ms");
    
    // Before optimization: ~2000-3000ms
    // After optimization: ~1000-1500ms (50% faster)
}
```

### 5. Measure Full Test Suite Impact (30min)
- Run full test suite with baseline config
- Run full test suite with optimized config
- Calculate percentage improvement

## Acceptance Criteria
- [ ] PostgreSQL container configured with test-optimized settings
- [ ] tmpfs volume configured for database storage
- [ ] Safety warnings and guards documented in code
- [ ] Benchmark shows 30-50% performance improvement
- [ ] Full test suite runs 20-30% faster
- [ ] All tests pass with new configuration
- [ ] Documentation updated with safety warnings

## Performance Metrics
- **Baseline**: Standard PostgreSQL configuration
  - 1000 inserts: ~2000-3000ms
  - Full test suite: ~200s
- **Optimized**: Test-specific configuration
  - 1000 inserts: ~1000-1500ms (50% faster)
  - Full test suite: ~140-160s (20-30% faster)
- **Expected Improvement**: 30-50% faster database operations

## Configuration Explanation

| Setting | Production | Test | Impact |
|---------|------------|------|--------|
| `fsync` | on | off | 2-5x faster writes |
| `synchronous_commit` | on | off | 2-3x faster commits |
| `full_page_writes` | on | off | Smaller WAL, faster writes |
| `max_connections` | 100 | 500 | Supports parallel tests |
| `shared_buffers` | 128MB | 256MB | Better caching |
| Storage | Disk | RAM (tmpfs) | 10-100x faster I/O |

## Safety Considerations

**Why this is safe for tests**:
- ✅ Test data is ephemeral (destroyed after tests)
- ✅ Container crashes don't lose real data
- ✅ Tests can be re-run on failure
- ✅ No external dependencies on test data

**Why this is UNSAFE for production**:
- ❌ Power loss = data loss
- ❌ Crash = corrupted database
- ❌ No recovery guarantees
- ❌ Violates ACID properties

**Guardrails**:
- Compiler directive prevents production builds
- Clear warning comments in code
- Runtime environment check
- Documentation emphasizes test-only usage

## Testing
```bash
# Verify PostgreSQL settings
docker exec <postgres-container> psql -U postgres -c "SHOW fsync;"
docker exec <postgres-container> psql -U postgres -c "SHOW synchronous_commit;"
docker exec <postgres-container> psql -U postgres -c "SHOW max_connections;"

# Verify tmpfs mount
docker exec <postgres-container> df -h | grep tmpfs

# Run performance benchmark
dotnet test --filter Benchmark_InsertPerformance --logger "console;verbosity=detailed"
```

## Potential Issues
- **Increased memory usage**: tmpfs uses RAM
  - ✅ 1GB allocation is reasonable for tests
  - ✅ Monitor CI runner memory usage
- **Container startup failure**: Insufficient memory
  - ✅ Reduce tmpfs size to 512MB if needed
- **Flaky tests**: Race conditions more visible with faster operations
  - ✅ Actually beneficial - exposes concurrency bugs early

## References
- Research: Section 4.2, 4.3
- PostgreSQL Docs: https://www.postgresql.org/docs/current/runtime-config-wal.html
- PostgreSQL Performance: https://wiki.postgresql.org/wiki/Tuning_Your_PostgreSQL_Server

## Dependencies
- Works with TEST-01 (IntegrationTestFixture)
- Synergizes with TEST-05 (transaction cleanup)

## Estimated Effort
3 hours

## Priority
P3 - Medium Impact, Low Effort (20-30% database operation improvement)

**Comment by DegrassiAaron** [comment]
## Closing as Optional - Marginal Gains, Foundation Sufficient

**Decision**: Closing as OPTIONAL (cost-benefit analysis suggests deferring)

### Analysis

**Proposed Optimizations**:
1. Disable fsync, synchronous_commit (test-only Postgres config)
2. Use tmpfs storage for database (RAM disk)
3. Increase shared_buffers, max_connections

**Expected Impact**:
- Database operations: 30-50% faster
- Test execution: 20-30% overall improvement (database is not bottleneck)

### Cost-Benefit Assessment

**Benefits**:
- Faster database writes (~2-5x)
- Better caching (256MB shared_buffers)
- Marginal test suite improvement (~20-30%)

**Costs**:
- Complexity: Medium (additional configuration)
- Safety concerns: Must ensure test-only (never production)
- Maintenance: Ongoing documentation and safety guards
- Risk: Complexity increase for marginal gains

**Current Performance**: Already acceptable
- Container startup: ~15s (shared)
- Test execution: ~50-100ms per test
- Full suite: Acceptable duration

### Decision Rationale

**Foundation is Sufficient**:
- ✅ CollectionFixture: 80-90% of gains achieved
- ✅ Parallelization: CPU optimized  
- ✅ Alpine images: Download optimized
- ✅ 100% test pass rate

**TEST-07 is Fine-Tuning**:
- Marginal improvement (20-30% ops, not overall)
- Complexity increase
- Diminishing returns
- Not worth effort at this stage

### When to Revisit

**Re-open if**:
- Database operations become bottleneck (profiling shows >50% time in DB)
- Test suite exceeds acceptable duration (e.g., >5 min)
- Performance regression detected

**Current**: Database is NOT the bottleneck (container startup and test logic dominate)

### Recommendation

**Close and monitor**:
- ✅ Current performance acceptable
- ✅ Foundation complete and solid
- ⏳ Implement only if profiling shows DB bottleneck

---

**Status**: OPTIONAL - Not needed at this time

Closes #615
